0856-Score-of-Parentheses.py

```diff
@@ -0,0 +1,49 @@
+'''
+Given a balanced parentheses string S, compute the score of the string based on the following rule:
+
+    () has score 1
+    AB has score A + B, where A and B are balanced parentheses strings.
+    (A) has score 2 * A, where A is a balanced parentheses string.
+
+ 
+
+Example 1:
+
+Input: "()"
+Output: 1
+
+Example 2:
+
+Input: "(())"
+Output: 2
+
+Example 3:
+
+Input: "()()"
+Output: 2
+
+Example 4:
+
+Input: "(()(()))"
+Output: 6
+
+ 
+
+Note:
+
+    S is a balanced parentheses string, containing only ( and ).
+    2 <= S.length <= 50
+'''
+# Time Complexity: O(n)
+# Space Complexity: O(1)
+class Solution:
+    def scoreOfParentheses(self, S: str) -> int:
+        res = bal = 0
+        for idx, s in enumerate(S):
+            if s == '(':
+                bal += 1
+            else:
+                bal -= 1
+                if S[idx - 1] == '(':
+                    res += 1 << bal
```

Comment:
can you please explain?

Reply:
keeping track of bal (balance) as to which parentheses had just come '(' or ')' and increasing decreasing its value accordingly.
And then whenever we come across '()' in S then we update the result by adding the value `res += 2**bal` as `bal` represents the exterior parentheses that has this core '()', which we can do by shifting 1 bit to the left i.e. `res += 1 << bal`
so basically `res += 2**bal` is replaced with `res += 1 << bal`